### PR TITLE
[api] readState support to non-rolldpos mode

### PR DESCRIPTION
**Pull request by envestcc**
# Description

`STANDALONE` and `NOOP` is easy to use than `ROLLDPOS` in writing e2etest, and `readState` is used to checking things in testing. But it's only supported in ROLLDPOS.

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [x] Code refactor or improvement

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: api/coreservice.go [hunk 1]
```diff
@@ -929,28 +929,30 @@ func (core *coreService) readState(ctx context.Context, p protocol.Protocol, hei
...
-	tipEpochNum := rp.GetEpochNum(tipHeight)
 	if height != "" {
 		inputHeight, err := strconv.ParseUint(height, 0, 64)
 		if err != nil {
 			return nil, uint64(0), err
 		}
-		inputEpochNum := rp.GetEpochNum(inputHeight)
-		if inputEpochNum < tipEpochNum {
+		rp := rolldpos.FindProtocol(core.registry)
+		if rp != nil {
```
Comment (dustinxie) [inline api/coreservice.go:938]: what if `rp == nil`, still cannot read?

File: api/coreservice.go [hunk 2]
```diff
@@ -929,28 +929,30 @@ func (core *coreService) readState(ctx context.Context, p protocol.Protocol, hei
...
+			tipEpochNum := rp.GetEpochNum(tipHeight)
+			if height != "" {
```
Comment (dustinxie) [inline api/coreservice.go:940]: no need, L932 already checked

File: api/coreservice.go (hunk 1 above)
Comment (envestcc) [reply api/coreservice.go:938]: it will go to L949 / L957 to read

File: api/coreservice.go (hunk 2 above)
Comment (envestcc) [reply api/coreservice.go:940]: indeed

File: api/coreservice.go [hunk 3]
```diff
@@ -929,28 +929,28 @@ func (core *coreService) readState(ctx context.Context, p protocol.Protocol, hei
...
 			return nil, uint64(0), err
 		}
-		inputEpochNum := rp.GetEpochNum(inputHeight)
-		if inputEpochNum < tipEpochNum {
+		rp := rolldpos.FindProtocol(core.registry)
+		if rp != nil {
+			tipEpochNum := rp.GetEpochNum(tipHeight)
+			inputEpochNum := rp.GetEpochNum(inputHeight)
+			if inputEpochNum < tipEpochNum {
+				inputHeight = rp.GetEpochHeight(inputEpochNum)
```
Comment (Liuhaai) [inline api/coreservice.go:942]: can't understand why the height of state is changed in rp, but the changes are  acceptable for me compared to the prev

**Review by Liuhaai — APPROVED** [review]

**Review by CoderZhi — APPROVED** [review]

**Pull request merged**